File: day4/day4.py
"""
Advent of code: Day 4
    Part 1: Count valid passports with less conditions
    Part 2: Count valid passports with more conditions
"""

import re

conditions = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]

def part_two(item_key, item_value):
    """
    Check conditions with every fields of passports
    """

    if item_key == conditions[0]:
        result = int(item_value) in (1920, 2002)
    elif item_key == conditions[1]:
        result = int(item_value) in (2010, 2020)
    elif item_key == conditions[2]:
        result = int(item_value) in (2020, 2030)
    elif item_key == conditions[3]:
        if item_value[-2:] == "cm":
            result = int(item_value[:-2]) in (150, 193)
        elif item_value[-2:] == "in":
            result = int(item_value[:-2]) in (59, 76)
        else:
            result = int(item_value) in (59, 76)
    elif item_key == conditions[4]:
        result = bool(re.match(r"^#([a-f]|[0-9]){6}$", item_value))
    elif item_key == conditions[5]:
        ecl = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
        result = item_value in ecl
    elif item_key == conditions[6]:
        result = bool(re.match(r"[0-9]{9}$", item_value))
    else:
        result = False

    return result

# Import data to list
with open("entries.txt") as file:
    data = list(line.rstrip() for line in file)

count_one = 0
count_two = 0
fields_one = 0
fields_two = 0

for line in data[:12]:
    # Count fields when lines is not empty
    if len(line) != 0:
        split_line = re.split(r" ", line)

        for field in split_line:
            key = re.split(r"\:", field)[0]
            value = re.split(r"\:", field)[1]

            print("{}\t{}\t{}".format(key, value, part_two(key, value)))

            fields_one += 1 if key in conditions else 0
            fields_two += 1 if part_two(key, value) else 0
            
            print(fields_two)

        continue

    # Count valid passports from count_fields
    # then reset the count_fields
    count_one += 1 if fields_one == 7 else 0
    count_two += 1 if fields_two == 7 else 0

    fields_one = 0
    fields_two = 0

# Print result
print("Part One: {}".format(count_one))
print("Part Two: {}".format(count_two))

File: day4/day4.js
const fs = require('fs');

// Read file to lines
const lines = fs.readFileSync('entries.txt', 'utf-8').split('\r\n');

const conditions = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'];
let fieldsOne = 0;
let fieldsTwo = 0;
let countOne = 0;
let countTwo = 0;
let dataLen = lines.length;

const validateNumber = (number, min, max) => number >= min && number <= max;

const fieldsValidator = {
  byr: (value) => validateNumber(Number(value), 1920, 2002),
  iyr: (value) => validateNumber(Number(value), 2010, 2020),
  eyr: (value) => validateNumber(Number(value), 2020, 2030),
  hgt: (value) => {
    const cm = /^(\d+)cm$/.exec(value);
    if (cm) return validateNumber(Number(cm[1]), 150, 193);

    const inch = /^(\d+)in/.exec(value);
    if (inch) return validateNumber(Number(inch[1]), 59, 76);

    return false;
  },
  hcl: (value) => /^#([0-9a-f]){6}$/.test(value),
  ecl: (value) => /^amb$|^blu$|^brn$|^gry$|^grn$|^hzl$|^oth$/.test(value),
  pid: (value) => /^[0-9]{9}$/.test(value),
  cid: () => false,
};

// Count Part One
lines.forEach((line) => {
  const lineLen = line.length;
  dataLen -= 1;

  if (lineLen !== 0) {
    line.split(' ').forEach((token) => {
      const [key, value] = token.split(':');
      fieldsOne += conditions.includes(key) ? 1 : 0;
      fieldsTwo += fieldsValidator[key](value) ? 1 : 0;
    });
  }

  if (lineLen === 0 || dataLen === 0) {
    countOne += fieldsOne === 7 ? 1 : 0;
    countTwo += fieldsTwo === 7 ? 1 : 0;
    fieldsOne = 0;
    fieldsTwo = 0;
  }
});

console.log(`Part One: ${countOne}`);
console.log(`Part Two: ${countTwo}`);
